test/parallel/test-stream-map.js

```diff
@@ -33,18 +72,69 @@ const { setTimeout } = require('timers/promises');
 }
 
 {
-  // Map works on asynchronous streams with a asynchronous mapper
-  const stream = Readable.from([1, 2, 3, 4, 5]).map(async (x) => {
+  // Does not care about data events
+  const source = new Readable({
+    read() {
+      this.push(Uint8Array.from([1]));
+      this.push(Uint8Array.from([2]));
+      this.push(null);
+    }
+  });
+  setImmediate(() => stream.emit('data', Uint8Array.from([1])));
+  const stream = source.map(async ([x]) => {
     return x + x;
   }).map((x) => x + x);
-  const result = [4, 8, 12, 16, 20];
+  const result = [4, 8];
   (async () => {
     for await (const item of stream) {
       assert.strictEqual(item, result.shift());
     }
   })().then(common.mustCall());
 }
 
+{
+  // Emitting an error during `map`
+  const stream = Readable.from([1, 2, 3, 4, 5]).map(async (x) => {
+    if (x === 3) {
+      stream.emit('error', new Error('boom'));
+    }
+    return x + x;
+  });
+  assert.rejects(
+    stream.map((x) => x + x).toArray(),
+    /boom/,
```

Comment:
We could check the error by reference here and below.
```suggestion
  const error = new Error();
  const stream = Readable.from([1, 2, 3, 4, 5]).map(async (x) => {
    if (x === 3) {
      stream.emit('error', error);
    }
    return x + x;
  });
  assert.rejects(
    stream.map((x) => x + x).toArray(),
    error,
```

Reply:
I'm not sure that's actually a guarantee we necessarily want to make in this case (that is - that if we emit we don't wrap the error in any way) - the guarantee currently tested is much smaller (that we include the message).

If you feel strongly it's a guarantee we _should_ make - I'll change it (we should also probably update the docs)

Reply:
If the current implementation does reject with the same error object, and in the future we decide to wrap this object in some other object, wouldn't you agree this would be a semver-major change? That's why I would say testing for it would make sense.

The proposal says we should `Await` at each step, so I would think that implies the same rejected object is returned. Maybe we should make another test that verifies that when using a non-`Error` objects, everything still behaves. Non-blocking suggestion.

Reply:
> wouldn't you agree this would be a semver-major change? 

Not on an experimental API - once it graduates 100% absolutely.

> The proposal says we should Await at each step, so I would think that implies the same rejected object is returned. 

Yes but note that in this case the user is explicitly `emit`ing the error event rather than throwing an error - I agree that for throwing we 100% should align.

----

> Non-blocking suggestion.

Please keep making suggestions discussing these things is important!